File: solutions/java/split-second-stopwatch/1/src/main/java/SplitSecondStopwatch.java
import java.util.List;
import java.util.ArrayList;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;

public class SplitSecondStopwatch {
    private static enum State {
        READY, RUNNING, STOPPED;
    }

    private final DateTimeFormatter fmt = DateTimeFormatter.ofPattern("HH:mm:ss");
    private State state = State.READY;
    private LocalTime lap = LocalTime.MIDNIGHT;
    private List<LocalTime> laps = new ArrayList<>();

    public void start() {
        if (state == State.RUNNING)
            throw new IllegalStateException("cannot start an already running stopwatch");
        state = State.RUNNING;
    }

    public void stop() {
        if (state != State.RUNNING)
            throw new IllegalStateException("cannot stop a stopwatch that is not running");
        state = State.STOPPED;
    }

    public void reset() {
        if (state != State.STOPPED)
            throw new IllegalStateException("cannot reset a stopwatch that is not stopped");
        state = State.READY;
        lap = LocalTime.MIDNIGHT;
        laps = new ArrayList<>();
    }

    public void lap() {
        if (state != State.RUNNING)
            throw new IllegalStateException("cannot lap a stopwatch that is not running");
        laps.add(lap);
        lap = LocalTime.ofSecondOfDay(0);
    }

    public String state() {
        return switch(state) {
        case READY -> "ready";
        case RUNNING -> "running";
        case STOPPED -> "stopped";
        };
    }

    public String currentLap() {
        return lap.format(fmt);
    }

    public String total() {
        return laps.stream()
            .reduce(lap, (acc, el) -> acc.plusSeconds(el.toSecondOfDay()))
            .format(fmt);
    }

    public List<String> previousLaps() {
        return laps.stream()
            .map(lap_ -> lap_.format(fmt))
            .toList();
    }

    public void advanceTime(String timeString) {
        if (state == State.RUNNING) {
            lap = lap.plusSeconds(LocalTime.parse(timeString).toSecondOfDay());
        }
    }
}

File: solutions/csharp/split-second-stopwatch/1/SplitSecondStopwatch.cs
public enum StopwatchState
{
    Ready,
    Running,
    Stopped
}

public class SplitSecondStopwatch(TimeProvider time)
{
	private TimeProvider _time = time;
	private List<TimeSpan> _previousLaps = new List<TimeSpan>();
	private DateTimeOffset? _lapStart;
	private TimeSpan? _elapsed;
	
    public StopwatchState State { get; private set; }
    public TimeSpan CurrentLap => 
		State switch
		{
				StopwatchState.Ready => TimeSpan.Zero,
				StopwatchState.Stopped => _elapsed.Value,
				StopwatchState.Running => (_time.GetUtcNow() - _lapStart.Value) + _elapsed.Value,
				_ => throw new NotImplementedException()
		};
    public TimeSpan Total =>
        _previousLaps.Aggregate(TimeSpan.Zero, (agg, ts) => agg + ts.Duration()) + CurrentLap;
    public IReadOnlyCollection<TimeSpan> PreviousLaps => _previousLaps.AsReadOnly();

    public void Start()
    {
		if(State == StopwatchState.Running) throw new InvalidOperationException();
        if(State == StopwatchState.Ready) _elapsed = new TimeSpan();
		_lapStart = _time.GetUtcNow();
        State = StopwatchState.Running;
    }

    public void Stop()
    {
        if(State != StopwatchState.Running) throw new InvalidOperationException();
        _elapsed = _time.GetUtcNow() - _lapStart.Value;
        _lapStart = null;
        State = StopwatchState.Stopped;
    }

    public void Reset()
    {
        if(State != StopwatchState.Stopped) throw new InvalidOperationException();
		_lapStart = null;
		_elapsed = null;
		_previousLaps.Clear();
		State = StopwatchState.Ready;
    }

    public void Lap()
    {
        if(State != StopwatchState.Running) throw new InvalidOperationException();
        _previousLaps.Add(CurrentLap);
        _lapStart = _time.GetUtcNow();
        _elapsed = new TimeSpan();
    }
}